libbeat/paths/paths.go

```diff
@@ -0,0 +1,143 @@
+// Package libbeat.paths provides a common way to handle paths
+// configuration for all Beats.
+//
+// Currently the following paths are defined:
+//
+// path.home - It’s the default folder for everything that doesn't fit in
+// the categories below
+//
+// path.data - Contains things that are expected to change often during normal
+// operations (“registry” files, UUID file, etc.)
+//
+// path.config - Configuration files and Elasticsearch template default location
+//
+// These settings can be set via the configuration file or via command line flags.
+// The CLI flags overwrite the configuration file options.
+//
+// Use the Resolve function to resolve files to their absolute paths. For example,
+// to look for a file in the config path:
+//
+// cfgfilePath := paths.Resolve(paths.Config, "beat.yml"
+package paths
+
+import (
+	"flag"
+	"fmt"
+	"os"
+	"path/filepath"
+
+	"github.com/elastic/beats/libbeat/logp"
+)
+
+var (
+	homePath   = flag.String("path.home", "", "Home path")
+	configPath = flag.String("path.config", "", "Configuration path")
+	dataPath   = flag.String("path.data", "", "Data path")
+)
+
+type Path struct {
+	Home   string
+	Config string
+	Data   string
+}
+
+// FileType is an enumeration type representing the file types.
+// Currently existing file types are: Home, Config, Data
+type FileType string
+
+const (
+	Home   FileType = "home"
+	Config FileType = "config"
+	Data   FileType = "data"
+)
+
+var paths Path
+
+// InitPaths sets the default paths in the configuration based on CLI flags,
+// configuration file and default values. It also tries to create the data
+// path with mode 0755 and returns an error on failure.
+func InitPaths(cfg *Path) error {
```

Comment:
My preference is to declare this as a method on a `*Path` receiver. Then make a singleton `Path` somewhere. So instead of having a function that affects global state you have singleton object with methods on it. This gives users of the class a bit more freedom in their implementation. For example, if I wanted to decouple my class from the global state, I could inject a `*Path` into my class making testing easier.